### PR TITLE
feat(database): support cell-level selection

### PR DESCRIPTION
**Pull request by zqran**
https://github.com/toeverything/blocksuite/assets/15389209/e07a2fdc-d36c-45ae-95b3-d484c56a56d8

### CODE REVIEW

**Comment by zqran** [comment]
> Hi @zqran, i think it's better to save the select ranges rather than single cell to support this feature. For more, it can achieve multi-selcet cells easily.

Thanks for your suggestion, it's helpful, I'll refactor it.

If you have other suggestions, it would be great to see all kinds of feedback.

**Comment by blurname** [comment]
> > Hi @zqran, i think it's better to save the select ranges rather than single cell to support this feature. For more, it can achieve multi-selcet cells easily.
> 
> Thanks for your suggestion, it's helpful, I'll refactor it.
> 
> If you have other suggestions, it would be great to see all kinds of feedback.

Thanks, if you counter any question about this, welcome to share. I recently implement a small feature like database in my work

**Comment by zqran** [comment]
> Thanks, if you counter any question about this, welcome to share. I recently implement a small feature like database in my work

Refactoring done. 🤝

File: packages/blocks/src/__internal__/utils/types.ts [hunk 1]
```diff
@@ -28,13 +28,24 @@ export interface EditingState {
...
-export type DatabaseTableState = {
-  type: DatabaseTableStateType;
+export type DatabaseTableViewRowStateType = 'select' | 'clear' | 'click';
+export type DatabaseTableViewRowState = {
+  type: DatabaseTableViewRowStateType;
   databaseId?: string;
   rowIds?: string[];
 };
 
+export type DatabaseTableViewCellStateType = 'select' | 'clear' | 'edit';
```
Comment (blurname) [inline packages/blocks/src/__internal__/utils/types.ts:38]: 1. if `Currently only supports single cell selection`, you can make coords just `coords:[CellCoord]` and left TODO there.
2. `DatabaseTableViewCellState` should be divided into 3 types for type narrowing and precise business logic.

examle:
```ts
type CellCoord = {
  rowIndex: number;
  cellIndex: number;
};

type DatabaseTableViewCellSelectState = {
  type: 'select'
  databaseId: string
  coords: [CellCoord]
}
type DatabaseTableViewCellClickState = {
  type: 'click'
  databaseId: string
  coords: [CellCoord]
}
type DatabaseTableViewCellClearState = {
  type: 'clear'
}
```

Comment (blurname) [reply packages/blocks/src/__internal__/utils/types.ts:38]: Then, you can remove redundant `?.`

File: packages/blocks/src/database-block/table/components/selection/utils.ts [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+export function getCellCoord(
+  target: CellCoord | HTMLElement,
+  databaseId: string,
+  key: string
+) {
+  const rowsContainer = getRowsContainer(databaseId);
+  const cellRects = getAllCellsRect(rowsContainer);
+
+  const rowsCount = cellRects.length;
+  const columnsCount = cellRects[0].length;
```
Comment (blurname) [inline packages/blocks/src/database-block/table/components/selection/utils.ts:98]: typo: cellsCount

File: packages/blocks/src/database-block/table/components/selection/cell-selection.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    return {
+      left: left - containerRect.left,
+      top: top - containerRect.top,
+      height,
+      width,
+    };
+  };
+
+  override render() {
+    const { left, top, height, width, display } = this._getStyles();
```
Comment (doodlewind) [inline packages/blocks/src/database-block/table/components/selection/cell-selection.ts:74]: The `display` here can be missing

**Review by doodlewind — APPROVED** [review]
Sorry for the delay, I think this feat is pretty much self-contained, thanks!

**Pull request merged**